### PR TITLE
Fix USDT run_time test targeting all pids

### PR DESCRIPTION
**Pull request by jordalgo**
This creates a new pid namespace (and mounts proc) to run bpftrace inside of this namespace.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Review by danobi — COMMENTED** [review]
The issue in the regular (non-embedded) CI is that it looks like `unshare` is not present in the nix flake config. Should be fairly easy to add it there: https://github.com/iovisor/bpftrace/blob/12c463e8ff71bb1423243cbc3d23a7c4d59d5b8b/flake.nix#L110

The other issue in the embedded CI kinda looks like a namespacing issue (guessing that's what the `[unknown]` bit is). I'm suspecting it'd be more reliable to add pid namespace awareness to the runtime test harness itself rather than having tests shell out to `unshare`. Cuz it requires a bit of care to run the BEFORE progs in the same pidns as bpftrace. And it looks like the infra already has to have some awareness.

Could use a marker directive like existing `WILL_FAIL`. That way other tests could easily reuse this. I see a lot of value for the other existing ones that walk procfs.

**Comment by jordalgo** [comment]
@danobi Let me know what you think of this first draft

**Review by danobi — COMMENTED** [review]
Should `AFTER` scripts run in the pid ns too? Might be ok to omit until actually necessary.

But docs should be updated with `NEW_PIDNS` as well as whatever we land on for `AFTER` behavior

File: tests/runtime/engine/runner.py [hunk 1]
```diff
@@ -196,10 +198,53 @@ def run_test(test):
...
+                    while waited <= test.timeout:
+                        get_unshare_children = ["ps", "--ppid", str(unshare_out.pid), "--no-headers", "-o", "pid"]
+                        unshare_children = subprocess.run(get_unshare_children,
+                                                          check=False,
+                                                          stdout=subprocess.PIPE,
+                                                          stderr=subprocess.PIPE)
+
+                        if unshare_children.returncode == 0 and unshare_children.stdout:
+                            lines = [line.decode('utf-8') for line in unshare_children.stdout.splitlines()]
+                            lines = sorted((line.strip() for line in lines))
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply tests/runtime/engine/runner.py:232]: Good call. Although at this point there really should only be one child.

**Comment by jordalgo** [comment]
> Should AFTER scripts run in the pid ns too? Might be ok to omit until actually necessary.

That's a good call. I think they should. I'll make that update.

File: tests/runtime/engine/runner.py [hunk 2]
```diff
@@ -136,6 +138,31 @@ def __get_bpffeature():
...
+                children = subprocess.run(run_cmd,
+                                          check=False,
+                                          stdout=subprocess.PIPE,
+                                          stderr=subprocess.PIPE)
+                if children.returncode == 0 and children.stdout:
+                    lines = [line.decode('utf-8') for line in children.stdout.splitlines()]
+                    if (condition(lines)):
+                        return True
+                else:
+                    print(f"__wait_for_children error: {children.stderr}")
```
Comment (danobi) [inline tests/runtime/engine/runner.py:157]:
```suggestion
                    print(fail(f"__wait_for_children error: {children.stderr}"))
```

File: tests/runtime/engine/runner.py [hunk 3]
```diff
@@ -196,42 +232,64 @@ def run_test(test):
...
+                    # Use the first BEFORE as the first process in the new pid namespace
+                    unshare_out = subprocess.Popen(["unshare", "--fork", "--pid", "--mount-proc", "-r", "--kill-child"] + test.befores[0].split(),
+                                                   start_new_session=True, universal_newlines=True,
+                                                   stdout=subprocess.PIPE,
+                                                   stderr=subprocess.STDOUT)
+
+                    def has_children(lines):
+                        if len(lines) == 0:
+                            return False
+                        nsenter.extend(["nsenter", "-p", "-m", "-t", lines[0].strip()])
```
Comment (danobi) [inline tests/runtime/engine/runner.py:250]: A bit confusing that the "conditional" callback mutates state in the caller

**Comment by danobi** [comment]
Ah didn't notice PR was in draft. Feel free to ignore comments while you debug

File: tests/runtime/engine/runner.py (hunk 3 above)
Comment (jordalgo) [reply tests/runtime/engine/runner.py:250]: I know. I didn't love this either. Maybe I could have this function return the lines if the conditional is true.

**Comment by jordalgo** [comment]
> Ah didn't notice PR was in draft. Feel free to ignore comments while you debug

No, please continue. These comments are helpful. I also still can't figure out why the Embedded builds are failing.

File: tests/runtime/engine/runner.py (hunk 2 above)
Comment (jordalgo) [reply tests/runtime/engine/runner.py:157]: This is just for debugging. I don't think we want it to have the fail colors.

**Comment by jordalgo** [comment]
@danobi This is ready for a proper review

File: tests/runtime/engine/runner.py [hunk 4]
```diff
@@ -162,7 +187,16 @@ def run_test(test):
         after = None
         after_output = None
         cleanup = None
+        # This is only populated if the NEW_PIDNS directivce is set
```
Comment (danobi) [inline tests/runtime/engine/runner.py:190]:
```suggestion
        # This is only populated if the NEW_PIDNS directive is set
```

File: tests/runtime/engine/runner.py [hunk 5]
```diff
@@ -162,7 +187,16 @@ def run_test(test):
...
+        # and is used to enter the newly created pid namespace for all BEFOREs,
+        # the prirmary RUN or PROG command, and the AFTER.
```
Comment (danobi) [inline tests/runtime/engine/runner.py:192]:
```suggestion
        # the primary RUN or PROG command, and the AFTER.
```

File: tests/runtime/engine/runner.py [hunk 6]
```diff
@@ -162,7 +187,16 @@ def run_test(test):
...
+        nsenter = []
         bpf_call = "[unknown]"
+
+        def get_pid_ns_cmd(cmd):
+            cwd = "." + os.getcwd() + "/"
+            return nsenter + [x.replace("./", cwd) for x in cmd.split()]
```
Comment (danobi) [inline tests/runtime/engine/runner.py:198]: Use abspath() like above? Also handles relative paths
```suggestion
            return nsenter + [os.path.abspath(x) for x in cmd.split()]
```

File: tests/runtime/engine/runner.py [hunk 7]
```diff
@@ -137,6 +139,29 @@ def __get_bpffeature():
         bpffeature["jiffies64"] = output.find("jiffies64: yes") != -1
         return bpffeature
 
+    
+    @staticmethod
+    def __wait_for_children(parent_pid, timeout, ps_format, condition):
+        with open(os.devnull, 'w') as dn:
```
Comment (danobi) [inline tests/runtime/engine/runner.py:145]: Unused?

File: tests/runtime/engine/runner.py [hunk 8]
```diff
@@ -196,42 +232,64 @@ def run_test(test):
...
+                        return True
+
+                    if not Runner.__wait_for_children(unshare_out.pid, test.timeout, "pid", has_children):
+                        raise TimeoutError(f'Timed out waiting create a new PID namespace')
+
+                    # This is the only one we need to add to befores as killing the
+                    # unshare process will kill the whole process subtree with "--kill-child"
+                    befores.append(unshare_out)
+                    for before in test.befores[1:]:
+                        child_name = os.path.basename(before.strip().split()[-1])[:15]
```
[the remark this answers is not in the document]
Comment (danobi) [reply tests/runtime/engine/runner.py:257]: Ah I see. It's assuming there are no arguments passed to the BEFORE program. For example with this diff things break:

```diff
diff --git a/tests/runtime/usdt b/tests/runtime/usdt
index 72f1da3c..5eb1b03f 100644
--- a/tests/runtime/usdt
+++ b/tests/runtime/usdt
@@ -90,7 +90,7 @@ NEW_PIDNS

 NAME usdt probes - attach to fully specified library probe by pid
 RUN {{BPFTRACE}} -e 'usdt:./testlibs/libusdt_tp.so:tracetestlib:lib_probe_1 { printf("here\n" ); exit(); }' -p $(pidof usdt_lib)
-BEFORE ./testprogs/usdt_lib
+BEFORE ./testprogs/usdt_lib la
 EXPECT here
 TIMEOUT 5
 REQUIRES ./testprogs/usdt_lib should_not_skip
```

**Comment by jordalgo** [comment]
> Ah I see. It's assuming there are no arguments passed to the BEFORE program. For example with this diff things break:

Good catch. Will fix.

File: tests/runtime/engine/runner.py (hunk 7 above)
Comment (jordalgo) [reply tests/runtime/engine/runner.py:145]: I stole this from the original code. TBH, I'm not sure what it's doing but I was afraid to just remove it.

**Review by danobi — APPROVED** [review]

**Pull request merged**